### ISSUE TITLE
The LLaMA implementation by Keras Hub exhibits significant deviations in accuracy compared to the standard implementation (Hugging Face).

### ISSUE THREAD

**Issue by pass-lin** [keras-team/keras-hub#1993]
```python
import os

os.environ["HF_ENDPOINT"] = "https://hf-mirror.com"#chinese huggingface mirror source
os.environ["KERAS_BACKEND"] = "torch"
os.environ["CUDA_VISIBLE_DEVICES"] = "1"
model_name = 'NousResearch/Meta-Llama-3.1-8B'
import torch
from transformers import AutoTokenizer,AutoModelForCausalLM,AutoConfig
import keras
hf_model = AutoModelForCausalLM.from_pretrained(model_name,
                                                device_map="cuda:0",
                                             torch_dtype=torch.bfloat16, 
                                             _attn_implementation = 'eager',
                                             trust_remote_code=False).eval()
import keras_hub
keras.config.set_dtype_policy('bfloat16')
keras_model = keras_hub.models.Llama3CausalLM.from_preset('hf://'+model_name)

tokenizer = AutoTokenizer.from_pretrained(model_name)
input_ids,mask = tokenizer('计算量决定了网络执行时间的长短，参数量决定了占用显存的量').values()
input_ids = keras.ops.expand_dims(input_ids,0)
mask = keras.ops.expand_dims(mask,0)

x1 = hf_model.forward(input_ids,attention_mask=mask)
x2 = keras_model([mask,input_ids])

error = keras.ops.abs(x1.logits-x2)

error = keras.ops.abs(x1.logits-x2)

print(keras.ops.max(error))
print(keras.ops.min(error))
print(keras.ops.mean(error))
print(keras.ops.std(error))

print(keras.ops.max(error,-1))
print(keras.ops.min(error,-1))
print(keras.ops.mean(error,-1))
print(keras.ops.std(error,-1))
```
The output is
···
tensor(3.2188, device='cuda:0', dtype=torch.bfloat16, grad_fn=<MaxBackward1>)
tensor(0., device='cuda:0', dtype=torch.bfloat16, grad_fn=<MinBackward1>)
tensor(0.2441, device='cuda:0', dtype=torch.bfloat16,
       grad_fn=<ToCopyBackward0>)
tensor(0.2129, device='cuda:0', dtype=torch.bfloat16, grad_fn=<StdBackward0>)
tensor([[0.5938, 0.4062, 2.0938, 1.0781, 1.2188, 2.4062, 2.2812, 1.5625, 1.5234,
         1.3750, 1.4844, 2.9531, 2.3281, 1.7344, 2.4062, 2.1875, 2.4062, 3.2188,
         1.6953, 1.6250, 1.7969, 1.5078]], device='cuda:0',
       dtype=torch.bfloat16, grad_fn=<AmaxBackward0>)
tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]],
       device='cuda:0', dtype=torch.bfloat16, grad_fn=<AminBackward0>)
tensor([[0.1060, 0.0669, 0.3340, 0.1436, 0.1621, 0.3320, 0.2617, 0.2236, 0.2246,
         0.2090, 0.2422, 0.2490, 0.2930, 0.2637, 0.2500, 0.3066, 0.3574, 0.3340,
         0.2676, 0.2598, 0.2344, 0.2461]], device='cuda:0',
       dtype=torch.bfloat16, grad_fn=<ToCopyBackward0>)
tensor([[0.0679, 0.0510, 0.2578, 0.1138, 0.1270, 0.2617, 0.2080, 0.1719, 0.1768,
         0.1631, 0.1855, 0.2090, 0.2285, 0.2061, 0.2002, 0.2432, 0.2812, 0.2715,
         0.2090, 0.2031, 0.1816, 0.1904]], device='cuda:0',
       dtype=torch.bfloat16, grad_fn=<StdBackward0>)

···
The implementations of standard implementation (Hugging Face) and Keras have significant accuracy differences in logits, both of which are based on the PyTorch backend to avoid framework-specific errors to a certain extent. In practical use, it's also observed that the LLaMA implementation by Keras Hub tends to have repetitive decoding more easily, while the implementations by HF and VLLM are less prone to repetitive decoding.
Is it necessary to fix this precision difference?

**Comment by pass-lin** [comment]
Upon further attempts, I found that not only does the issue exist under bf16, but a similar magnitude of error occurs under fp32 as well. Typically, we consider an error tolerance of 1e-5 or below to be acceptable under fp32, but here the error is significantly higher than that number. Therefore, the implementation of the llama model may have a considerable margin of error.

```python
import os

os.environ["HF_ENDPOINT"] = "https://hf-mirror.com"#chinese huggingface mirror source
os.environ["KERAS_BACKEND"] = "torch"
#os.environ["CUDA_VISIBLE_DEVICES"] = "1"
model_name = 'NousResearch/Meta-Llama-3.1-8B'
import torch
from transformers import AutoTokenizer,AutoModelForCausalLM,AutoConfig
import keras
hf_model = AutoModelForCausalLM.from_pretrained(model_name,
                                                device_map="cuda:1",
                                             torch_dtype=torch.float32, 
                                             _attn_implementation = 'eager',
                                             trust_remote_code=False).eval()
import keras_hub
#keras.config.set_dtype_policy('bfloat16')
keras_model = keras_hub.models.Llama3CausalLM.from_preset('hf://'+model_name)

tokenizer = AutoTokenizer.from_pretrained(model_name)
input_ids,mask = tokenizer('计算量决定了网络执行时间的长短，参数量决定了占用显存的量').values()
input_ids = keras.ops.expand_dims(input_ids,0)
mask = keras.ops.expand_dims(mask,0)

x1 = hf_model.forward(input_ids.cuda("cuda:1"),attention_mask=mask.cuda("cuda:1"))
x2 = keras_model([mask,input_ids])

error = keras.ops.abs(x1.logits.cpu()-x2.cpu())

print(keras.ops.max(error))
print(keras.ops.min(error))
print(keras.ops.mean(error))
print(keras.ops.std(error))

print(keras.ops.max(error,-1))
print(keras.ops.min(error,-1))
print(keras.ops.mean(error,-1))
print(keras.ops.std(error,-1))
```

```python
import os

os.environ["HF_ENDPOINT"] = "https://hf-mirror.com"#chinese huggingface mirror source
os.environ["KERAS_BACKEND"] = "torch"
#os.environ["CUDA_VISIBLE_DEVICES"] = "1"
model_name = 'NousResearch/Meta-Llama-3.1-8B'
import torch
from transformers import AutoTokenizer,AutoModelForCausalLM,AutoConfig
import keras
hf_model = AutoModelForCausalLM.from_pretrained(model_name,
                                                device_map="cuda:1",
                                             torch_dtype=torch.float32, 
                                             _attn_implementation = 'eager',
                                             trust_remote_code=False).eval()
import keras_hub
#keras.config.set_dtype_policy('bfloat16')
keras_model = keras_hub.models.Llama3CausalLM.from_preset('hf://'+model_name)

tokenizer = AutoTokenizer.from_pretrained(model_name)
input_ids,mask = tokenizer('计算量决定了网络执行时间的长短，参数量决定了占用显存的量').values()
input_ids = keras.ops.expand_dims(input_ids,0)
mask = keras.ops.expand_dims(mask,0)

x1 = hf_model.forward(input_ids.cuda("cuda:1"),attention_mask=mask.cuda("cuda:1"))
x2 = keras_model([mask,input_ids])

error = keras.ops.abs(x1.logits.cpu()-x2.cpu())

print(keras.ops.max(error))
print(keras.ops.min(error))
print(keras.ops.mean(error))
print(keras.ops.std(error))

print(keras.ops.max(error,-1))
print(keras.ops.min(error,-1))
print(keras.ops.mean(error,-1))
print(keras.ops.std(error,-1))
```

**Comment by mattdangerw** [comment]
Thanks! Will take a look.

**Comment by pass-lin** [comment]
> Thanks! Will take a look.

Hello, have you found out what might have caused it?

**Comment by pass-lin** [comment]
> Thanks! Will take a look.

Hello, have you found out what might have caused it?

**Comment by mehtamansi29** [comment]
Hi @pass-lin -

We have debug the issue and Here in Llama3CausalLM class model in keras_hub  Llama3Backbone and Llama3CausalLMPreprocessor is used. 
In the Llama3Backbone seems pending to be implemented([Llama3Backbone](https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/models/llama3/llama3_backbone.py#L8))
As here Llama3Backbone is subclass of LlamaBackbone. So due to LlamaBackbone class properties we can get the result but with accuracy deviations. 
We'll debug more and work on Llama3Backbone implementations soon. Thanks..!!!

**Comment by pass-lin** [comment]
@mehtamansi29 Hello, has this bug been fixed in the newly released 0.19?

**Comment by maitry63** [comment]
Hi @pass-lin , Thank you for reporting this! 
I tried this with `TinyLlama-1.1B-Chat-v1.0` model & got this output -

```python
Max Error:  tensor(2.6591, device='cuda:0', grad_fn=<MaxBackward1>)
Min Error:  tensor(0., device='cuda:0', grad_fn=<MinBackward1>)
Mean Error: tensor(0.1381, device='cuda:0', grad_fn=<MeanBackward1>)
Std Error:  tensor(0.1614, device='cuda:0', grad_fn=<StdBackward0>)

tensor([[2.6591, 2.2401, 1.2671, 1.1791, 1.5769, 1.0757, 0.8284, 1.0266, 1.9641,
         1.8035, 0.5838, 0.8220, 0.6646, 0.7869, 0.4712, 0.6301, 0.8153, 0.6626,
         0.3628, 0.7426, 0.6006, 0.4415, 0.5404, 0.6447, 1.5016, 0.5885, 0.3965,
         0.7020, 0.4640, 0.3507, 0.5487, 0.4820, 0.4326, 0.3324, 0.2639, 0.3652,
         0.5238, 0.4496, 0.4110, 0.2584, 0.4678, 0.6352]], device='cuda:0',
       grad_fn=<AmaxBackward0>)
tensor([[7.6294e-06, 1.3590e-05, 9.5367e-07, 9.0599e-06, 3.8147e-06, 9.0003e-06,
         7.7188e-06, 2.9802e-07, 3.3379e-06, 2.8610e-06, 1.4305e-06, 4.7684e-07,
         7.8678e-06, 9.0599e-06, 5.1260e-06, 4.7684e-07, 5.6028e-06, 1.2875e-05,
         7.6294e-06, 1.2398e-05, 4.7684e-06, 7.1526e-07, 9.5367e-06, 2.2650e-06,
         4.1723e-06, 2.4438e-06, 1.0431e-06, 2.0266e-06, 8.5831e-06, 3.0398e-06,
         2.1458e-06, 5.4836e-06, 3.3379e-06, 2.1458e-06, 1.9073e-06, 5.0068e-06,
         2.1458e-06, 9.5367e-07, 8.3447e-07, 0.0000e+00, 2.3842e-06, 1.4305e-06]],
       device='cuda:0', grad_fn=<AminBackward0>)
tensor([[0.5554, 0.4032, 0.1949, 0.1770, 0.2790, 0.1807, 0.1536, 0.1864, 0.2650,
         0.2776, 0.1025, 0.1238, 0.1541, 0.1254, 0.0830, 0.0992, 0.1333, 0.1049,
         0.0732, 0.1237, 0.0995, 0.0942, 0.1212, 0.1411, 0.2627, 0.0999, 0.0665,
         0.1115, 0.0982, 0.0861, 0.1057, 0.0847, 0.0827, 0.0639, 0.0458, 0.0697,
         0.0840, 0.0608, 0.0611, 0.0474, 0.0412, 0.0782]], device='cuda:0',
       grad_fn=<MeanBackward1>)
tensor([[0.3971, 0.3091, 0.1544, 0.1372, 0.2152, 0.1336, 0.1180, 0.1441, 0.2037,
         0.2154, 0.0805, 0.0950, 0.1066, 0.0991, 0.0633, 0.0756, 0.1026, 0.0785,
         0.0551, 0.0937, 0.0759, 0.0654, 0.0819, 0.0974, 0.1926, 0.0773, 0.0514,
         0.0876, 0.0699, 0.0565, 0.0786, 0.0655, 0.0600, 0.0478, 0.0353, 0.0535,
         0.0652, 0.0471, 0.0454, 0.0360, 0.0325, 0.0621]], device='cuda:0',
       grad_fn=<StdBackward0>)
```
check this [gist](https://colab.sandbox.google.com/gist/maitry63/f73ded1c56c6e41a01bd302001de0588/segofrmer_train.ipynb#scrollTo=INPJ6Csna-Ro).

With the latest updates, Keras Hub’s LLaMA 3 implementation still produces logit prediction scores that drift by up to ~2.66 compared to the official Hugging Face.  So, the bug still persists.

**Comment by maitry63** [comment]
Hi @pass-lin , After further investigating and testing using `meta-llama/Llama-3.2-1B` instead of TinyLlama (As already uses the default RoPE theta (10,000)). The comparison between the Hugging Face and Keras outputs gave:

```text
Max Error  : 0.081624
Mean Error : 0.008168
```
Updated [gist](https://colab.sandbox.google.com/gist/maitry63/f73ded1c56c6e41a01bd302001de0588/segofrmer_train.ipynb#scrollTo=8G2Z3SxV-pV_).

These values are much smaller than the previous error. This shows that the large logit mismatch is no longer present.

The [conversion script ](https://github.com/keras-team/keras-hub/blob/master/tools/checkpoint_conversion/convert_llama3_checkpoints.py) for `Llama3` checks whether the two models produce almost the same output using:

```python
np.testing.assert_allclose(
    keras_hub_logits,
    hf_output_logits,
    atol=1e-4,
)
```

This means the expected difference between the Hugging Face and Keras logits is less than `0.0001`. The current errors are still larger than this limit, so the test does not pass yet. However, the difference is now much smaller than before, which shows that the previous large mismatch has been resolved. 
The small remaining difference is likely due to normal numerical differences between the Hugging Face and Keras implementations and does not indicate the same issue as before.
